runtime/flang/fpcvt.c

```diff
@@ -806,32 +828,61 @@ __fortio_ecvt(double value, int ndigit, int *decpt, int *sign, int round)
     /* Is there anything else? */
   }
 
-  ieee_v.d = value;
-  fexp = ieee_v.v.e - 1023;
-  if (fexp == 1024) {
-    if (ieee_v.v.hm == 0 && ieee_v.v.lm == 0) {
-      strcpy(tmp, "Inf");
-      *sign = ieee_v.v.s;
-      *decpt = 0;
-      return tmp;
-    } else {
-      strcpy(tmp, "NaN");
-      *sign = 0;
-      *decpt = 0;
-      return tmp;
+  if (!is_quad) {
+    ieee_v.d = value;
+    fexp = ieee_v.v.e - EXPONENT_BIAS_D;
+    if (fexp == EXPONENT_BIAS_D + 1) {
+      if (ieee_v.v.hm == 0 && ieee_v.v.lm == 0) {
+        strcpy(tmp, "Inf");
+        *sign = ieee_v.v.s;
+        *decpt = 0;
+        return tmp;
+      } else {
+        strcpy(tmp, "NaN");
+        *sign = 0;
+        *decpt = 0;
+        return tmp;
+      }
     }
-  }
 
-  *sign = ieee_v.v.s;
-  ieee_v.v.s = 0;
-  value = ieee_v.d;
+    *sign = ieee_v.v.s;
+    ieee_v.v.s = 0;
+    value = ieee_v.d;
+  } else {
+    ieeeq_v.q = lvalue;
+    fexp = ieeeq_v.v.e - EXPONENT_BIAS_Q;
+    if (fexp == EXPONENT_BIAS_Q + 1) {
+      if (ieeeq_v.v.lm[SUBSCRIPT_2] == 0 && ieeeq_v.v.lm[1] == 0 &&
+          ieeeq_v.v.lm[0] == 0 && ieeeq_v.v.hm == 0) {
+        if (width < INFINITY_STR_LEN + ieeeq_v.v.s)
+          strcpy(tmp, "Inf");
+        else
+          strcpy(tmp, "Infinity");
+        *sign = ieeeq_v.v.s;
+        *decpt = 0;
+        return tmp;
+      } else {
+        strcpy(tmp, "NaN");
+        *sign = 0;
+        *decpt = 0;
+        return tmp;
+      }
+    }
+
+    *sign = ieeeq_v.v.s;
+    ieeeq_v.v.s = 0;
+    lvalue = ieeeq_v.q;
+  }
 
   /* For compatible mode, round '5' away from zero */
   /* Compatible rounding, or compatible in number of good bits??? */
 
   if (round == FIO_COMPATIBLE) {
-      writefmt(fmt, ndigit, 'E');
-      j = sprintf(tmp, fmt, value);
+      writefmt(fmt, ndigit, 'E', is_quad);
+      if (!is_quad)
+        j = sprintf(tmp, fmt, value);
+      else
+        j = sprintf(tmp, fmt, lvalue);
```

Comment:
use of snprintf is recommended over sprintf

Reply:
...and this also goes to the many occurrences of this below.

Reply:
Fixed.